feat(zone-ingress-summary-view): use dropdown to select config format

schogges opened the pull request:
This is part of the efforts to allow the user using the preferred format of showing the config: `structured` vs `yaml` and adds a dropdown menu to select between `structured` and `yaml` in the `ZoneIngressSummaryView`.

![Screenshot 2025-01-17 at 18 31 58](https://github.com/user-attachments/assets/5c31392a-6095-4320-ad6e-6ae4238ac383)

Part of #3283

schogges commented on packages/kuma-gui/cypress/support/step_definitions/index.ts:
```diff
@@ -150,6 +150,11 @@ When(/^I click the "(.*)" element(?: and select "(.*)")?$/, (selector: string, v
   }
 })
 
+When(/^I select "(.*)" from the "(.*)" element$/, (value: string, selector: string) => {
+  $(selector).click()
+  $(`[data-testid="select-item-${value}"] button`).click()
+})
+
```
Note: this is very specific to how [KSelect](https://kongponents.konghq.com/components/select.html) works

johncowen replied:
oh yeah, I get this is a little awkward, although we've usually just done two clicks for this to avoid anything project specific in the step definitions

https://github.com/kumahq/kuma-gui/blob/09a9318112583014e162ac6d64397f03e0c7dbdd/packages/kuma-gui/features/mesh/dataplanes/Index.feature#L137-L139

schogges replied:
Ah, right 💡 so the step definition is a little easier to use and if the `KSelect` changes how it works we would only need to change it in one place 🤔

johncowen replied:
I'd be super wary of baking this sort of stuff into the steps, especially `data-testid` CSS selectors like this. I think at the least in this case I'm not sure what would happen if we had two KSelect components on the page with the same value.

I would guess thats easy to fix, but I'd still much prefer to try and avoid baking in application specific `data-testid` CSS selectors into the steps.

schogges replied:
Good point, I'll remove this 👍

schogges commented:
Flake

![image](https://github.com/user-attachments/assets/9b045595-1ee4-4c60-a33d-ad8e9d54ae1a)

johncowen commented on packages/kuma-gui/src/app/zone-ingresses/views/ZoneIngressSummaryView.vue:
```diff
@@ -52,110 +53,142 @@
-            <div
-              class="stack-with-borders"
+            <XLayout
+              type="stack"
```
Did we lose the border here intentionally?

schogges replied:
Oh, nice catch 👍 added it back 🙂

johncowen commented:
Something looks off to me here:

![Screenshot 2025-01-20 at 14 56 20](https://github.com/user-attachments/assets/51c36605-0d38-49d1-9af6-cacd1349c885)

compare the bottom padding with say a dataplane summary:

![Screenshot 2025-01-20 at 14 56 31](https://github.com/user-attachments/assets/05a90abc-64d2-4f8f-9283-cac7de2d6ff1)

lmk!

schogges commented:
> Something looks off to me here:
> 
> ![Screenshot 2025-01-20 at 14 56 20](https://private-user-images.githubusercontent.com/554604/404920818-51c36605-0d38-49d1-9af6-cacd1349c885.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczODU1MDIsIm5iZiI6MTczNzM4NTIwMiwicGF0aCI6Ii81NTQ2MDQvNDA0OTIwODE4LTUxYzM2NjA1LTBkMzgtNDlkMS05YWY2LWNhY2QxMzQ5Yzg4NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyMFQxNTAwMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00OTNjNjgwZDUyOWNhNjFkMWY0YjRiNzZmMTQ0NWI4NzdlZGM3OWU3NjhjY2VhMjk4MmMzODc0NGE1MGFhM2Q3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CxQcudCZjDBMNwYzUQpmXllRDkMA43yVVHeN630LaJw)
> 
> compare the bottom padding with say a dataplane summary:
> 
> ![Screenshot 2025-01-20 at 14 56 31](https://private-user-images.githubusercontent.com/554604/404920952-05a90abc-64d2-4f8f-9283-cac7de2d6ff1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczODU1MDIsIm5iZiI6MTczNzM4NTIwMiwicGF0aCI6Ii81NTQ2MDQvNDA0OTIwOTUyLTA1YTkwYWJjLTY0ZDItNGY4Zi05MjgzLWNhYzdkZTJkNmZmMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyMFQxNTAwMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hOTQ4YzZhNmU1NjJjYzRlODg3MWUyODIwMzA4ZWYxZTVhZjdjZTE4NGUyOTE0NThhZGVjZTRjODcyMjBjZGU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.xtF94ljo-S_uRaaAcFaRa9ojLWc7F_L6gD4GJxO0qrk)
> 
> lmk!

Argh 😄  It's because I switched to move `XLayout` and the `size` prop defaults to `normal` which adds a margin to each child element. I added another option `none`.

johncowen commented:
> Argh 😄 It's because I moved to XLayout and the size prop defaults to normal which adds a margin to each child element. I added another option none.

Ah gotcha, umm lets be a little bit more cautious with this, especially since its XLayout which is currently in like "try-out period".

🤔 An option of `none` would be the same as just not adding the `XLayout` at all right? So I don't think we should add that as an option, at least at the moment. Do you follow what I mean.

I think if anything `size="small"` is what you want, but I'm even a little cautious of doing:

```yaml
  <XLayout
    type="stack"
    size="small"
    class="stack-with-borders"
    data-testid="structured-view"
  >
```

This applies two lots of stacking styles, `stack-with-borders` and `XLayout`'s `stack`. I think this, coupled with my opinion that this layout is really a two column table with row scoped `th`'s, make me think we should just avoid changing this in this PR. 

Umm I suppose that aside, I'd suggest one of two options here:

1. Leave as is. Keep using `<div class="stack-with-borders">` as before. The feature/PR is adding the select functionality, and whilst it's great that we are upgrading things as we go, we've hit some complication here so we should be cautious.
2. If we want to upgrade to XLayout here/now, I'd go with `<XLayout type="stack" size="small" class="stack-with-borders">` but I really don't like mixing the two approaches to stacking just because we want borders so this would be my least fave option.

---

FYI: As its vaguely related to this point, I'd like to delete DefinitionCard entirely soon-ish.

schogges commented:
> > Argh 😄 It's because I moved to XLayout and the size prop defaults to normal which adds a margin to each child element. I added another option none.
> 
> Ah gotcha, umm lets be a little bit more cautious with this, especially since its XLayout which is currently in like "try-out period".
> 
> 🤔 An option of `none` would be the same as just not adding the `XLayout` at all right? So I don't think we should add that as an option, at least at the moment. Do you follow what I mean.
> 
> I think if anything `size="small"` is what you want, but I'm even a little cautious of doing:
> 
> ```yaml
>   <XLayout
>     type="stack"
>     size="small"
>     class="stack-with-borders"
>     data-testid="structured-view"
>   >
> ```
> 
> This applies two lots of stacking styles, `stack-with-borders` and `XLayout`'s `stack`. I think this, coupled with my opinion that this layout is really a two column table with row scoped `th`'s, make me think we should just avoid changing this in this PR.
> 
> Umm I suppose that aside, I'd suggest one of two options here:
> 
> 1. Leave as is. Keep using `<div class="stack-with-borders">` as before. The feature/PR is adding the select functionality, and whilst it's great that we are upgrading things as we go, we've hit some complication here so we should be cautious.
> 2. If we want to upgrade to XLayout here/now, I'd go with `<XLayout type="stack" size="small" class="stack-with-borders">` but I really don't like mixing the two approaches to stacking just because we want borders so this would be my least fave option.
> 
> FYI: As its vaguely related to this point, I'd like to delete DefinitionCard entirely soon-ish.

Aight gotcha 👍 I just though we are moving everything to `XLayout`, even `class="stack-with-borders"` at some point 🤔

johncowen commented:
> Aight gotcha 👍 I just though we are moving everything to XLayout, even class="stack-with-borders" at some point 🤔

Yeah we kinda are, but just not totally sure how yet, `borders` isn't a layout thing, so shouldn't be part of `XLayout`. This is partly why I think it should probably use a `Table` with row scoped `th`s. These `structured` views look and use the exact same design/CSS as our list tables. So I'm pretty sure I'd want them to somehow use the same CSS.

It would be great if our listing tables could also use row scoped `th`s (it would solve a few minor warts we have), just not quite sure how we can do that yet (if at all)

schogges commented:
> > Aight gotcha 👍 I just though we are moving everything to XLayout, even class="stack-with-borders" at some point 🤔
> 
> Yeah we kinda are, but just not totally sure how yet. `borders` isn't a layout thing, so shouldn't be part of `XLayout`. This is partly why I think it should probably use a `Table` with row scoped `th`s. These `structured` views look and use the exact same design/CSS as our list tables. So I'm pretty sure I'd want them to somehow use the same CSS.
> 
> It would be great if our listing tables could also use row scoped `th`s (it would solve a few minor warts we have), just not quite sure how we can do that yet (if at all)

Fair point 🙂 so we probably add another component? From my point of view that's the difference of a table vs list. A table has columns where each row shows the data related to that column. A list is just not that strict 🤔

johncowen commented:
> Fair point 🙂 so we probably add another component?

Not totally sure yet, this is why I think we should stick to adding the format menu here.

Ideally I want an `XTable` that we use here and in _our_ lists. But not quite sure yet if thats going to be possible because `KTable` won't let you add `th`'s in the `tbody` or add `scope` attributes to `th`s.

> A table has columns where each row shows the data related to that column.

I'd say this layout has columns. Two columns, the key and the value. The key column is left aligned the value column is right aligned. If you think about it like that they look exactly the same as our list tables.

---

Anyway, this is double-y why I think we just add the format menu here. I appreciate we are going through and upgrading things elsewhere as we go, I just think that this needs a tiny bit more thought first.
